Current

LouisJenkinsCS opened the pull request:
@qtcwt TGraph.hpp contains new graph information, GraphRecovery.hpp contains new graph test.

Commit:
Working on TGraph transformation

Commit:
Making steady progress, need to configure graph to keep around constant fixed size

Commit:
Have iniitalization of graph and metric obtainment

Commit:
Have sketched out overall idea to follow for tomorrow

Commit:
Making steady progress, need to debug segfault.

Commit:
Revising for paper deadline

Commit:
Working now, have all 4 operations for transient

Commit:
Running code on node2x20a

Commit:
Added hasher function and turned off optimizations to better be able to debug

Commit:
Fixed issue with std::unordered_set not finding things, working on memory std::bad_alloc exceptions

Commit:
Pushing out update with larger mean edges per vertex

qtcwt commented on src/rideables/TGraph.hpp:
```diff
@@ -161,171 +258,240 @@ class TGraph : public RGraph{
+            std::vector<int> vec(meanEdgesPerVertex);
+            for (size_t i = 0; i < meanEdgesPerVertex * 100 / vertexLoad; i++) {
+                int u = uniformVertex(vertexGen);
+                while (u == i) {
```
Why we discard u if it equals i?

LouisJenkinsCS replied:
We don't allow self-loops.

qtcwt commented on src/rideables/TGraph.hpp:
```diff
@@ -161,171 +258,240 @@ class TGraph : public RGraph{
+                retval = false;
+            }
+
+            std::reverse(vec.begin(), vec.end());
```
I think `std::reverse` is O(n), and is inside locks. Can we just do `for(auto u=vec.rbegin();u!=vec.rend();u++)` rather than reversing it and then iterating.
Another way, if you want to try new things, C++20 provides `views::reverse`. https://en.cppreference.com/w/cpp/ranges/reverse_view

LouisJenkinsCS replied:
Great point! Makes sense there is a reverse iterator built-in that would be more efficient. Also aren't we compiling for C++17?

qtcwt replied:
Oops that's true and I doubt the current version of gcc on 2x20a fully supports c++20. Let's just stay with the former solution for now..

qtcwt reviewed (commented):
Just dumping my preliminary reviews real quick; more to come after PST lunch.

qtcwt commented on src/main.cpp:
```diff
@@ -3,6 +3,7 @@
 #include <thread>
 #include <random>
 #include <time.h>
+#include <ratio>
```
something no longer in use?

LouisJenkinsCS replied:
Yeah, I originally tried keeping the `vertexLoad` as a real number, but C++ does not allow either `double` or `float` to be passed to as template parameter arguments due to it being unimplemented, and I wanted to use `ratio` instead. I decided to fall back on using an integer to denote the load instead; I'll remove this.

qtcwt commented on src/rideables/TGraph.hpp:
```diff
@@ -161,171 +258,240 @@ class TGraph : public RGraph{
+        void lock(size_t idx) {
+            std::atomic<bool>& lck = vertexLocks[idx];
+            bool expect = false;
+            while (lck.load() == true || lck.compare_exchange_strong(expect, true) == false) {
```
Any reason why we implement busy waiting locks by our own rather than just using `std::mutex`? I think `std::mutex::lock` yields CPUs and waits for the signal if the lock is occupied.

LouisJenkinsCS replied:
`std::atomic<bool>` has a size of 1 byte, and `std::mutex` has a size of 40 bytes; we maintain one lock per vertex, so shrinking the space requirement by 40x is ideal here, especially since the locks are held in transient memory.

qtcwt replied:
I’m worried about the overhead of being busy-waiting. Also, 1 Byte suffers from false sharing. Honestly, 40B*1M (Orkut has ~3M vertices which is close to our 1M) doesn’t sound big to me, especially given that we already use 4B as sequence number and 8B as idxToVertex indexing for each vertex, let alone transient tVertex which is at least 40B; I don’t think the size of mutex would become the space bottleneck, but it’s possible to be the performance bottleneck.

qtcwt commented on src/rideables/TGraph.hpp:
```diff
@@ -65,90 +79,173 @@ class TGraph : public RGraph{
+                        k = verticesRNG(gen);
+                    }
+                    if (idxToVertex[k] != nullptr) {
+                        Relation *in = new Relation(i, k, -1);
```
What if relation(i,k) already exists?

LouisJenkinsCS replied:
Then nothing happens; the `std::unordered_set` rejects the insertion. Hence, it skips over vertices that are null, but also does not insert redundant Relations... this does leak memory though, now that you've pointed it out though, so I should check to see if the insertion succeeded or not.

qtcwt replied:
Yes, memory leak was the concern. Thanks.

qtcwt commented on src/rideables/TGraph.hpp:
```diff
@@ -65,90 +79,173 @@ class TGraph : public RGraph{
+                    }
+                    if (idxToVertex[k] != nullptr) {
+                        Relation *in = new Relation(i, k, -1);
+                        Relation *out = new Relation(i, k, -1);
```
Should this be `new Relation(k, i, -1)`, if the semantic here is to add two edges of both directions? If this isn't the case, i.e., if it just adds a single edge and update on both src and dest, why do we create two relation to represent the same edge?

LouisJenkinsCS replied:
> why do we create two relation to represent the same edge?

To prevent issues where I run into a segmentation fault in cases where I accidentally free a shared `Relation` early and to simplify the design (originally I used a `std::shared_ptr but removed it once was running into bugs and decided to take it one step at a time).

qtcwt replied:
An alternative could be always to free relations in src of vertices only, and use dest just as auxiliary indexing. Am I right?

qtcwt commented on src/rideables/TGraph.hpp:
```diff
@@ -65,90 +79,173 @@ class TGraph : public RGraph{
             }
-            v->unlock();
+            Relation r(src, dest, -1);
+            retval = has_relation(source(src), &r);
```
I'm worried about the case when we are extending it to Montage Graph, we have to create a relation node for search, which is persistent. Perhaps we have to use `template< class K > iterator find( const K& x )` introduced in C++20, inevitably? https://en.cppreference.com/w/cpp/container/unordered_set/find

qtcwt replied:
Just tried locally, even gcc 10 doesn't support this overloaded find. We'll have to do some trick for Montage Graph..

LouisJenkinsCS replied:
It should be searching each element via the `RelationEqual` comparator defined above, so it should really just dereference the pointer. It shouldn't be a problem. It should work if the node being searched for is stack allocated, heap allocated, or NVRAM allocated, I believe.

qtcwt replied:
Allocating a relation node on NVM just for comparison sounds slow given the read latency of NVM... but you are right, stack allocation doesn’t have such problem.

qtcwt commented on src/rideables/TGraph.hpp:
```diff
@@ -161,171 +258,240 @@ class TGraph : public RGraph{
+            if (idxToVertex[src] != nullptr && idxToVertex[dest] != nullptr) {
+                Relation r(src, dest, -1);
+                remove_relation(source(src), &r);
+                remove_relation(destination(dest), &r);
```
A heads-up: if as I suggested above, finally we keep only one `Relation` node for an edge, and share it between `source(src)` and `destination(dest)`, then we should remove the pointer from both lists but just reclaim the `Relation` node once.

LouisJenkinsCS replied:
I'll consider it an optimization.

qtcwt commented on src/rideables/TGraph.hpp:
```diff
@@ -161,171 +258,240 @@ class TGraph : public RGraph{
                 for (int _vid : vertices) {
-                    idxToVertex[_vid]->lock();
+                    lock(_vid);
+                    if (idxToVertex[_vid] == nullptr && get_seq(vid) == seq) {
```
Is this only for debugging? It seems like `get_seq(vid) == seq` will not be evaluated at all, as the first condition is always false during correct execution.

LouisJenkinsCS replied:
Yup, it was for debugging; the bug itself was resolved when I found out that `std::unordered_set` requires a custom comparator when dealing with pointers. I can remove this code soon.

qtcwt commented on src/rideables/TGraph.hpp:
```diff
@@ -161,171 +258,240 @@ class TGraph : public RGraph{
-                            toRemoveList.push_back(r);
+                    Relation src(other, vid, -1);
+                    Relation dest(vid, other, -1);
+                    if (!has_relation(source(other), &src) && !has_relation(destination(other), &dest)) {
```
For full coverage while debugging, should the conditions be linked with _OR_ like `!has_relation(source(other), &src) || !has_relation(destination(other), &dest)`?

LouisJenkinsCS replied:
My fault for not catching this: It just has to find that _one_ of these relations exist because we extract vertices from both `source` and `destination` sets, it isn't necessary for it to be in both. We don't keep track of which set the vertex `other` belongs to, just that it was an edge in `source` or `destination`.

qtcwt commented on src/rideables/TGraph.hpp:
```diff
@@ -161,171 +258,240 @@ class TGraph : public RGraph{
-                for (Relation *r : garbageList) {
-                    delete r;
-                }
+                std::vector<Relation*> toDelete(source(vid).size() + destination(vid).size());
```
Can we just delete those relation nodes during iteration? Why do we need a vector to store them first?

LouisJenkinsCS replied:
See above.

qtcwt commented on src/rideables/TGraph.hpp:
```diff
@@ -65,90 +79,173 @@ class TGraph : public RGraph{
-                idxToVertex[i]->adjacency_list.clear();
-                idxToVertex[i]->dest_list.clear();
+            for (auto i = 0; i < numVertices; i++) {
+                std::vector<Relation*> toDelete(source(i).size() + destination(i).size());
```
Why do we need a vector to store these relation nodes? Can we just delete them during the iteration, or maybe I missed something?

LouisJenkinsCS replied:
I'm not sure if this is safe or not, and this is another thing I did to simplify the code to address other bugs and concerns, ruling out certain potential issues like this. I'm not sure if it is safe to delete an element while we are still iterating, as the `std::unordered_set` uses both a hash function and comparator function, so I'm not sure if, say, it decides to resize/shrink the map, it shifts around elements and ends up segfaulting when it reads recently freed memory. It is mostly implementation-defined behavior whether or not it is safe, and so I went with the safe option of first storing all `Relation` objects in a vector first prior to deleting them.

qtcwt replied:
Oh that’s right. Thanks for pointing it out.

qtcwt commented on src/rideables/TGraph.hpp:
```diff
@@ -65,90 +79,173 @@ class TGraph : public RGraph{
             for (int i = numVertices - 1; i >= 0; i--) {
-                idxToVertex[i]->seqNumber++;
-                idxToVertex[i]->unlock();
+                destroy(i);
```
Should we worry about `idxToVertex[i]` being `nullptr`?

LouisJenkinsCS replied:
We call `destroy(i)` which does this.

LouisJenkinsCS replied:
That's a great point; `clear()` is not actually used right now, at least not by the planned benchmark.

qtcwt replied:
Then we can leave it as is for now..

qtcwt commented on src/rideables/TGraph.hpp:
```diff
@@ -161,171 +258,240 @@ class TGraph : public RGraph{
+            return vertexSeqs[idx];
+        }
+
+        void destroy(size_t idx) {
```
`assert(idxToVertex[idx]!=nullptr)` for debugging?

LouisJenkinsCS replied:
Good point, that'd be a good assertion to add.

qtcwt commented on src/tests/GraphTest.hpp:
```diff
@@ -64,54 +69,49 @@ class GraphTest : public Test {
             gtc->interval = numeric_limits<double>::max();
+            auto stats = g->grab_stats();
+            std::apply(print_stats, stats);
+            insertionProb = 3300;
```
Make it more generic, like parameters or environment variables (passed during run time via `-d`)?

LouisJenkinsCS replied:
I'm unfamiliar with how to obtain this information. Is there an example somewhere?

qtcwt replied:
Yes. The former one is simple: just add four more parameters to the constructor of GraphTest. The latter is a little trickier: when you run `bin/main ... -dInsertProb=3300`, gtc will generate a record and you can get it by 
```
if(gtc->checkEnv("InsertProb"))
   insertionProb = atoi((gtc->getEnv("InsertProb")).c_str());
```

qtcwt commented on src/tests/GraphTest.hpp:
```diff
@@ -64,54 +69,49 @@ class GraphTest : public Test {
+                    g->add_edge(distv(gen_v), distv(gen_v), -1);
+                } else if (rng <= insertionProb + removalProb) {
+                    // std::cout << "rng(" << rng << ") is remove_any_edge <= " << insertionProb + removalProb << std::endl; 
+                    g->remove_edge(distv(gen_v), distv(gen_v));
```
The probability of successfully adding a new edge (i.e., P(succeeds in adding an edge | adding an edge) is higher than successfully removing an existing edge, I think. Say at the beginning each vertex has 128 edges on average and there're 1048576 vertex slots and 524288 existing vertices, then the probability of successfully adding an edge is 1/2*((524288-128)/1048575), while the probability of successfully removing an edge is 1/2*(128/1048575). The latter is significantly less than the former, which means the average number of edges are increasing during the execution.

LouisJenkinsCS replied:
:thinking:

roghnin replied:
FWIW, I do think this issue also appears in ChurnTest for maps and sets. It's normally not a problem when we set prefill the data structure to half full.

qtcwt replied:
Actually I think these are different: successfully adding and removing a node from map/set both have probability of 1/2 if the container is filled by half. But as I said, successfully adding and removing an edge have different probability: the former is 1/2*((524288-128)/1048575) but the latter is 1/2*(128/1048575). The problem is the that _edges for each vertex_ is not filled in half

qtcwt replied:
To be clear: filling edges by half means 1/2 of existing vertices (which are 1/2 of all vertex slots) neighbor every vertex.

qtcwt replied:
Just did a quick calculation: if in total there are 1048575 (1024*1024) vertices, on average half of them exist, and each vertex has 128 edges on average, then add_edge:remove_edge ratio has to be 1/4096 to keep the average number of edges of each vertex around 128.
Suppose `Vavg` is the average number of existing vertices, `Vall` the total number of vertex slots, `E` the average number of edges in each vertex. The goal is to get add_edge:remove_edge ratio `x`. Then the equation is `x(Vavg-E)/Vall=(1-x)E/Vall`. After simplification, `x=E/Vavg`. We can take only probability of `edge_ops` and `vertex_ops` as parameter during execution and calculate the ratio of add_edge:remove_edge (i.e., `x`) during initialization (add_vertex:remove_vertex is simply 1:1).

roghnin replied:
That's indeed too many edges given the vertex number. Let's go with a pre-set smaller ratio then.

roghnin replied:
Does this also mean that the possibility of _actually_ deleting/adding an edge is pretty low? Do you want to consider a smaller graph?

qtcwt replied:
I've pushed my solution in commit [59a9723](https://github.com/urcs-sync/Montage/pull/23/commits/59a9723967a627f41ca4d0c8b7ed5fc8813235d6). See what you think.

LouisJenkinsCS replied:
Here are some results from newer changes...

```
|V|=528594,|E|=68210682,average degree = 129.042,maximum degree = 327
|V|=304841,|E|=17752737,average degree = 58.2361,maximum degree = 5339
add_edge = 15 (0.00339982%)
, remove_edge = 96 (0.0217588%)
, add_vertex = 108668 (24.6301%)
, remove_vertex = 332421 (75.3447%)
```

Even with varying numbers of tasks, the ratio is about the same here. The average degree drops pretty sharply, here, which is due to `remove_vertex` succeeding 3x more often than `add_vertex`.

qtcwt replied:
Notice that the average degree stays stable when `add_edge:remove_edge=3:1`. I don't understand why but as an ad-hoc solution I just pushed it to the PR. Trying to find out where this magical 3:1 comes from for another 30 minutes and give up afterwards.

Commit:
count the time for writes-back and reclamation into the length of epochs, and report epoch growth.

Commit:
Copied TGraph.hpp, added tVertex class, and changed references to vertices to point to tVertex objects.

roghnin committed:
put some cout into gtc->verbose condition

Commit:
NVMGraph.hpp

Commit:
fixed syntax errors

Commit:
Got rid of false sharing by switching to mutex; performance is decent but having problems

qtcwt committed:
use reverse iterator rather than reversing entire vector

qtcwt committed:
combine three per vertex metadata into aligned struct

qtcwt committed:
pass edge_op:vertex_op as env var
calculate corrected ratio of edge add anf remove to keep #edges stable

qtcwt reviewed (commented):
I've incorporated some of my suggestions into the code in my commits [cd38f7b](https://github.com/urcs-sync/Montage/pull/23/commits/cd38f7be3a1aab999cfbad4d052b65746e2a6b34), [9374813](https://github.com/urcs-sync/Montage/pull/23/commits/937481309b94a283ffd4b101a72e3f324d45f22c), and [59a9723](https://github.com/urcs-sync/Montage/pull/23/commits/59a9723967a627f41ca4d0c8b7ed5fc8813235d6), and marked those incorporated suggestions as "solved." Those conversations haven't been marked "solved" are not done.
HEADSUP: although I did my best checking code several times before pushing, since I'm not able to run either on node1x10a due to missing `boost/historgam.hpp` or on my laptop due to large memory consumption, there might be stupid run-time bugs which however should be easy to fix; I deliberately make each commit for just one purpose to my best effort... I would fix them tomorrow morning if haven't when I wake up.

LouisJenkinsCS commented:
To install boost

```
git clone https://github.com/spack/spack.git
spack/bin/spack install boost
spack/bin/spack load boost
```

While doing the last command it may ask you to run a command; do that and run `spack load boost`.

Commit:
Updating assertion test condition

LouisJenkinsCS commented:
Pushed out change to the debugging test condition so it doesn't fail.

LouisJenkinsCS committed:
Merge pull request #1 from LouisJenkinsCS/NVMGraph

Copied TGraph.hpp, added tVertex class, and changed references to ver…

Commit:
Adding helper 'vertex' method and fixed NVMGraph

Commit:
Adding NVMGraph to main

Commit:
Running NVMGraph and MontageGraph

Commit:
Added a new test

Commit:
Testing MontageGraph

Commit:
Merge branch 'master' of github.com:LouisJenkinsCS/Montage

Commit:
Added op limit

Commit:
Merge branch 'master' of github.com:LouisJenkinsCS/Montage

Commit:
AllocTest is complete

Commit:
AllocTest now runs correctly with a single thread

qtcwt committed:
balance add_edge:remove_edge in ad-hoc way

Commit:
Working on MontageGraph

qtcwt commented on src/rideables/MontageGraph.hpp:
```diff
@@ -221,348 +290,405 @@ class MontageGraph : public RGraph, public Recoverable{
-                for (int _vid : vertices) {
-                    idxToVertex[_vid]->seqNumber++;
-                    idxToVertex[_vid]->unlock();
+                // Step 4: Release in reverse order
```
`end_op` before releasing locks; let's have a pair of curly braces from line 579 to 620

qtcwt committed:
end_op before unlock

qtcwt committed:
Merge branch 'master' of github.com:LouisJenkinsCS/Montage into louis

qtcwt committed:
bug in MontageGraph()

qtcwt committed:
remove unnecessary begin_op

roghnin committed:
Merge https://github.com/LouisJenkinsCS/Montage into graph_louis

roghnin committed:
moved alloc() out of critical section in add_edge()

Commit:
move allocation out of critical section in add_dege of MontageGraph and NVMGraph

qtcwt committed:
update run_graph.sh

qtcwt committed:
allocate one relation per edge
fix bug of add_edge and remove_edge return value
fix add_edge:remove_edge ratio bug
fix hash bug in MontageGraph
use Montage (T) rather than NVMGraph
